### PR TITLE
Bug 1822665: Find vm by name if UUID search fails

### PR DESCRIPTION
**Pull request by alexander-demichev**
This PR should fix the case when VM misses UUID and `Exists()` call fails to find VM. Failing to find VM causes the controller to try to create an existing VM, which leads to endless loop of failing tasks.
The proposed fix is to fallback to searching by name when UUID search fails.

### CODE REVIEW

**Comment by JoelSpeed** [comment]
> this looks good to me, i wonder if we could be using the errors.IsNotFound from apimachinery instead of writing our own?
> 
> https://github.com/kubernetes/apimachinery/blob/master/pkg/api/errors/errors.go#L480

As I've understood this, the not found errors here are from the vSphere API and not the K8s API. The one you've linked is very specifically formatted to the K8s API and so can't really be used with any other API I don't think

**Comment by elmiko** [comment]
@JoelSpeed got it, thanks!

i thought these errors were probably different than the client errors, but i wanted to ask to make sure.

File: pkg/controller/vsphere/session/session.go [hunk 1]
```diff
@@ -105,20 +105,20 @@ func GetOrCreate(
...
-		klog.V(3).Infof("Find template by instance uuid %v", ID)
-		ref, err := s.FindRefByInstanceUUID(ctx, ID)
-		if err != nil {
-			return nil, errors.Wrap(err, "error querying template by instance UUID")
-		}
-		if ref != nil {
-			return object.NewVirtualMachine(s.Client.Client, ref.Reference()), nil
-		}
+func (s *Session) FindVM(ctx context.Context, UUID, name string) (*object.VirtualMachine, error) {
+	if !isValidUUID(UUID) {
```
Comment (JoelSpeed) [inline pkg/controller/vsphere/session/session.go:109]: Looks like this logic was wrong originally, oops 😅

File: pkg/controller/vsphere/session/session.go [hunk 2]
```diff
@@ -105,20 +105,20 @@ func GetOrCreate(
...
+		klog.V(3).Infof("No valid UUID %v, trying to find by name", name)
```
Comment (JoelSpeed) [inline pkg/controller/vsphere/session/session.go:110]: Should this log message include the UUID given? 
```suggestion
		klog.V(3).Infof("Invalid UUID for VM %q: %s, trying to find by name", name, UUID)
```

File: pkg/controller/vsphere/session/session.go [hunk 3]
```diff
@@ -105,20 +105,20 @@ func GetOrCreate(
...
+		return s.FindVMByName(ctx, name)
 	}
-
-	klog.V(3).Infof("Find template by name %v", ID)
-	return s.findVMByName(ctx, ID)
+	klog.V(3).Infof("Find template by instance uuid %v", UUID)
```
Comment (JoelSpeed) [inline pkg/controller/vsphere/session/session.go:113]:
```suggestion
	klog.V(3).Infof("Find template by instance uuid: %s", UUID)
```

File: pkg/controller/vsphere/session/session.go [hunk 4]
```diff
@@ -105,20 +105,20 @@ func GetOrCreate(
...
+	ref, err := s.FindRefByInstanceUUID(ctx, UUID)
+	if ref != nil && err == nil {
+		return object.NewVirtualMachine(s.Client.Client, ref.Reference()), nil
+	}
+	if err != nil {
+		klog.V(3).Infof("Not found by UUID: %v, trying to find by name %v", err, name)
```
Comment (JoelSpeed) [inline pkg/controller/vsphere/session/session.go:119]:
```suggestion
		klog.V(3).Infof("Instance not found by UUID: %s, trying to find by name %q", err, name)
```

File: pkg/controller/vsphere/session/session.go [hunk 5]
```diff
@@ -138,14 +138,26 @@ func (s *Session) findRefByUUID(ctx context.Context, UUID string, findByInstance
 	return ref, nil
 }
 
-func (s *Session) findVMByName(ctx context.Context, ID string) (*object.VirtualMachine, error) {
+func (s *Session) FindVMByName(ctx context.Context, ID string) (*object.VirtualMachine, error) {
```
Comment (JoelSpeed) [inline pkg/controller/vsphere/session/session.go:141]: Does this need to be changed to an exported function?

**Comment by alexander-demichev** [comment]
@JoelSpeed All comments fixed

File: pkg/controller/vsphere/reconciler_test.go [hunk 6]
```diff
@@ -992,33 +992,15 @@ func TestDelete(t *testing.T) {
 }
 
 func TestCreate(t *testing.T) {
-	model, session, server := initSimulator(t)
-	defer model.Remove()
```
Comment (enxebre) [inline pkg/controller/vsphere/reconciler_test.go:996]: Is there any reason why this change is in this commit?

Comment (alexander-demichev) [reply pkg/controller/vsphere/reconciler_test.go]: Thanks for finding it. I don't remember why it's like this :D

File: pkg/controller/vsphere/reconciler_test.go [hunk 7]
```diff
@@ -899,7 +899,7 @@ func TestDelete(t *testing.T) {
 	password, _ := server.URL.User.Password()
 	namespace := "test"
 	vm := simulator.Map.Any("VirtualMachine").(*simulator.VirtualMachine)
-	instanceUUID := "instanceUUID"
+	instanceUUID := "a5764857-ae35-34dc-8f25-a9c9e73aa898"
```
Comment (JoelSpeed) [inline pkg/controller/vsphere/reconciler_test.go:902]: Was there not an extra test case here with an invalid UUID to test the behaviour of the fallback? Seems to have gone?

Comment (alexander-demichev) [reply pkg/controller/vsphere/reconciler_test.go:902]: Do you mean we should let invalid value stay here? Tests are passing in both cases anyway.

Comment (JoelSpeed) [reply pkg/controller/vsphere/reconciler_test.go:902]: I was wondering if there are any tests that actually cover the fallback to name? For that we would need an invalid ID right? But I couldn't see any tests where the ID was invalid, are there tests where the parse of the ID fails?

Comment (alexander-demichev) [reply pkg/controller/vsphere/reconciler_test.go:902]: What about this test https://github.com/openshift/machine-api-operator/pull/558/files#diff-1cda1db15f2622f564de1259e970d48eR161 ?

Comment (JoelSpeed) [reply pkg/controller/vsphere/reconciler_test.go:902]: I am blind 🤦  Ignore me :D
